File: CORE/PY/PitfallsAndErrors.py
"""
================================================================================
PitfallsAndErrors.py - Python Competitive Programming Traps & Debugging Guide
================================================================================
A comprehensive diagnostic handbook detailing common traps, subtle runtime
pitfalls, and performance bottlenecks in Python that cause Time Limit Exceeded (TLE),
Memory Limit Exceeded (MLE), Wrong Answer (WA), and Runtime Errors (RTE) on
Codeforces, LeetCode, and HackerRank.

TABLE OF CONTENTS:
--------------------------------------------------------------------------------
1.  [PIT-01] Recursion Limit Exceeded & Stack Overflow (RTE)
2.  [PIT-02] Codeforces Anti-Hash Collision Attacks (O(N^2) TLE in Dict/Set)
3.  [PIT-03] Floating-Point Precision Loss & Float Division Gotchas (WA)
4.  [PIT-04] O(N^2) String Concatenation in Loops (TLE)
5.  [PIT-05] The O(N) list.pop(0) / list.insert(0) Queue Trap (TLE)
6.  [PIT-06] The 2D Matrix Shallow Copy Trap (`[[0]*M]*N`) (WA)
7.  [PIT-07] Bitwise Operator Precedence Traps (WA)
8.  [PIT-08] Lexicographical Number Comparison Trap (WA)
9.  [PIT-09] Modifying Collections While Iterating (RTE / Skipped Elements)
10. [PIT-10] Slicing Inside Loops Causing Hidden O(K) Copies (TLE)
11. [PIT-11] Slow Standard I/O Overhead on 10^5+ Lines (TLE)
12. [PIT-12] Python's Round-Half-To-Even "Banker's Rounding" (WA)
================================================================================
"""

import sys
import math
import random
import time
from collections import deque


# ==============================================================================
# [PIT-01] Recursion Limit Exceeded & Stack Overflow (RTE)
# ==============================================================================
"""
Trap:
    Python's default recursion limit is 1,000 stack frames.
    Any deep DFS, tree traversal, or divide-and-conquer algorithm on graphs
    with N >= 1,000 will throw `RecursionError: maximum recursion depth exceeded`.

Why It Happens:
    Python maintains detailed frame objects for stack traces, so the interpreter
    guards against C-level stack overflow with a strict artificial limit.

How To Fix It:
    1. Increase the limit at the very top of your solution:
       `sys.setrecursionlimit(300000)`
    2. On platforms like Codeforces running on Windows, if the native C stack is
       exceeded, use an iterative DFS with an explicit list stack, or run your logic
       in a separate thread with a custom stack size:
       `threading.stack_size(1024 * 1024 * 128)` (128 MB stack).
"""

def demonstrate_recursion_fix():
    # Demonstrates safe recursion limit expansion
    old_limit = sys.getrecursionlimit()
    sys.setrecursionlimit(5000)
    
    def deep_sum(n):
        if n == 0:
            return 0
        return 1 + deep_sum(n - 1)
    
    result = deep_sum(2500)
    sys.setrecursionlimit(old_limit)  # Restore
    return result


# ==============================================================================
# [PIT-02] Codeforces Anti-Hash Collision Attacks (O(N^2) TLE in Dict/Set)
# ==============================================================================
"""
Trap:
    On Codeforces (and some competitive programming platforms), tests are generated
    specifically to exploit Python's deterministic built-in integer hash function:
        `hash(x) == x` (for small integers).
    Adversaries construct inputs where all values share identical hash buckets
    modulo the dictionary's table size, triggering continuous hash collisions.
    This degrades `dict` and `set` lookups from O(1) average to O(N) worst-case,
    causing an otherwise optimal O(N) algorithm to take O(N^2) and TLE!

Why It Happens:
    Python hashes integers directly to their integer values. The hash table capacity
    is always a power of 2, making collision generation trivial for problem setters.

How To Fix It:
    Use a randomized 64-bit salt XOR-ed with each key:
        `RANDOM_SALT = random.getrandbits(61)`
        `def safe_hash(x): return x ^ RANDOM_SALT`
    Wrap keys or use a custom dict class.
"""

RANDOM_SALT = random.getrandbits(61)

class SafeDict(dict):
    """Wrapper that prevents anti-hash collision attacks by salting keys."""
    def __setitem__(self, key, value):
        super().__setitem__(key ^ RANDOM_SALT, value)
    def __getitem__(self, key):
        return super().__getitem__(key ^ RANDOM_SALT)
    def __contains__(self, key):
        return super().__contains__(key ^ RANDOM_SALT)
    def get(self, key, default=None):
        return super().get(key ^ RANDOM_SALT, default)


# ==============================================================================
# [PIT-03] Floating-Point Precision Loss & Float Division Gotchas (WA)
# ==============================================================================
"""
Trap:
    1. Using `/` instead of `//` converts exact integers into 64-bit IEEE-754 floats.
       Floats only have 53 bits of mantissa (~15 to 17 decimal digits of precision).
       For 64-bit integers (e.g., 10^18), float division silently truncates precision:
           `int(10**18 / 3 * 3) != 10**18`
    2. Using `int(math.sqrt(x))` on large integers (e.g., 10^18) causes precision drift.

How To Fix It:
    - Always use integer floor division `//` when working with integer arithmetic.
    - Use `math.isqrt(x)` for exact integer square roots (Python 3.8+).
"""

def demonstrate_precision_pitfall():
    large_int = 10**18 + 7
    float_res = int((large_int // 10) * 10)  # Exact integer math
    return float_res


# ==============================================================================
# [PIT-04] O(N^2) String Concatenation in Loops (TLE)
# ==============================================================================
"""
Trap:
    Writing `s += char` inside a loop of size N.
    Strings in Python are immutable. Each `+=` allocates a new string in memory
    and copies all previous i characters, resulting in:
        1 + 2 + 3 + ... + N = O(N^2) operations.
    For N = 10^5, this takes ~15 seconds and causes instant TLE.

How To Fix It:
    Append characters or chunks to a list, then join them once at the end:
        `parts = []`
        `for char in stream: parts.append(char)`
        `result = "".join(parts)`  # Strictly O(N)
"""

def fast_string_build(n: int) -> str:
    """Builds a string of length N in O(N) time instead of O(N^2)."""
    parts = []
    for i in range(n):
        parts.append('a')
    return "".join(parts)


# ==============================================================================
# [PIT-05] The O(N) list.pop(0) / list.insert(0) Queue Trap (TLE)
# ==============================================================================
"""
Trap:
    Using a standard Python `list` as a queue via `q.pop(0)` or `q.insert(0, val)`.
    `list.pop(0)` removes the first element and shifts all remaining N - 1 elements
    one position to the left in memory, costing O(N) time per operation.
    In a BFS on 10^5 nodes, this results in O(N^2) = 10^10 operations (TLE).

How To Fix It:
    Always use `from collections import deque`.
    `dq.popleft()` and `dq.appendleft()` are strictly O(1).
"""

def demonstrate_queue_performance():
    dq = deque()
    for i in range(1000):
        dq.append(i)
    # Fast O(1) popleft
    return dq.popleft()


# ==============================================================================
# [PIT-06] The 2D Matrix Shallow Copy Trap (`[[0]*M]*N`) (WA)
# ==============================================================================
"""
Trap:
    Creating a 2D matrix using:
        `grid = [[0] * M] * N`
    This creates ONE single list `[0] * M` and duplicates its REFERENCE N times!
    Modifying `grid[0][0] = 5` will simultaneously set `grid[1][0] = 5`,
    `grid[2][0] = 5`, etc., corrupting the entire grid!

How To Fix It:
    Use a list comprehension to instantiate a distinct row list on every iteration:
        `grid = [[0] * M for _ in range(N)]`
"""

def create_safe_matrix(rows: int, cols: int, fill=0) -> list:
    """Instantiates a 2D grid where every row is an independent list in memory."""
    return [[fill] * cols for _ in range(rows)]


# ==============================================================================
# [PIT-07] Bitwise Operator Precedence Traps (WA)
# ==============================================================================
"""
Trap:
    Arithmetic operators (+, -) have HIGHER precedence than bitwise shifts (<<, >>).
    Comparison operators (==, !=, <) have HIGHER precedence than bitwise AND/OR (&, |).
    Examples:
        `1 << n - 1`   evaluates as `1 << (n - 1)`, NOT `(1 << n) - 1`!
        `x & 1 == 0`   evaluates as `x & (1 == 0)` which is `x & False` -> 0!
        `x & y == target` evaluates as `x & (y == target)`!

How To Fix It:
    ALWAYS enclose bitwise operations in explicit parentheses:
        `((1 << n) - 1)`
        `(x & 1) == 0`
        `(x & y) == target`
"""

def is_even_bitwise(x: int) -> bool:
    # Explicit parentheses are mandatory!
    return (x & 1) == 0

def mask_all_ones(n: int) -> int:
    # Generates (2^n - 1)
    return (1 << n) - 1


# ==============================================================================
# [PIT-08] Lexicographical Number Comparison Trap (WA)
# ==============================================================================
"""
Trap:
    Comparing numerical inputs while they are still stored as strings:
        `"100" < "20"` evaluates to `True`!
    Because Python compares strings character-by-character from left to right,
    and `'1' < '2'`.

How To Fix It:
    Always cast string inputs to integers before comparisons:
        `int("100") < int("20")` -> `False`
    If sorting strings by numeric value without casting the whole string:
        `arr.sort(key=int)`
"""

def sort_numeric_strings(arr: list) -> list:
    """Sorts string representations of numbers by their numeric value."""
    return sorted(arr, key=int)


# ==============================================================================
# [PIT-09] Modifying Collections While Iterating (RTE / Skipped Elements)
# ==============================================================================
"""
Trap:
    1. In sets / dicts: Mutating size during iteration raises:
       `RuntimeError: Set changed size during iteration`
    2. In lists: Removing an element while iterating shifts subsequent elements,
       causing the loop index to advance past the neighbor without inspecting it!

How To Fix It:
    - Iterate over a shallow copy of the collection:
      `for item in list(my_set):`
      `for key in list(my_dict.keys()):`
    - Or build a filtered container using a list comprehension:
      `arr = [x for x in arr if condition(x)]`
"""

def remove_negatives_safely(arr: list) -> list:
    """Safely filters items without mutation-skipping bugs."""
    return [x for x in arr if x >= 0]


# ==============================================================================
# [PIT-10] Slicing Inside Loops Causing Hidden O(K) Copies (TLE)
# ==============================================================================
"""
Trap:
    Inside a loop over length N, writing `window = arr[i:i+k]`.
    Python list slicing makes a SHALLOW COPY of the elements.
    If K is large (e.g., K = 50,000) and N = 100,000:
    The loop performs 100,000 copies of size 50,000 = 5 * 10^9 operations (TLE)!

How To Fix It:
    - Use index pointers instead of copying sublists.
    - Update running metrics (sum, Counter, hash) incrementally when entering
      and leaving the window in O(1) instead of re-slicing and re-aggregating.
"""

def running_window_sum(arr: list, k: int) -> list:
    """Maintains sliding window sum incrementally in O(1) per step with zero slicing."""
    curr = sum(arr[:k])
    res = [curr]
    for i in range(k, len(arr)):
        curr += arr[i] - arr[i - k]  # O(1) update
        res.append(curr)
    return res


# ==============================================================================
# [PIT-11] Slow Standard I/O Overhead on 10^5+ Lines (TLE)
# ==============================================================================
"""
Trap:
    Calling `input()` and `print()` 100,000+ times.
    `input()` performs single-character reads and strips newlines, while `print()`
    flushes and formats strings repeatedly, creating massive runtime overhead.

How To Fix It:
    Use fast I/O:
    ```python
    import sys
    # Read entire input at once:
    input_data = sys.stdin.read().split()
    
    # Fast output:
    sys.stdout.write("\\n".join(output_lines) + "\\n")
    ```
"""

def parse_all_integers(stream: str) -> list:
    """Simulates ultra-fast bulk integer parsing from stdin."""
    return list(map(int, stream.split()))


# ==============================================================================
# [PIT-12] Python's Round-Half-To-Even "Banker's Rounding" (WA)
# ==============================================================================
"""
Trap:
    In Python 3, `round(x)` implements Banker's Rounding (rounds half-integers
    to the nearest EVEN integer):
        `round(2.5) == 2`  (rounds DOWN to nearest even!)
        `round(3.5) == 4`  (rounds UP to nearest even!)
    Competitive programming problems often specify standard arithmetic rounding
    (round half UP away from zero, so 2.5 should become 3).

How To Fix It:
    For positive numbers: `int(x + 0.5)`
    Or using `decimal`:
    ```python
    from decimal import Decimal, ROUND_HALF_UP
    val = int(Decimal(str(x)).quantize(Decimal('1'), rounding=ROUND_HALF_UP))
    ```
"""

def standard_round_half_up(x: float) -> int:
    """Standard competitive programming half-up rounding (2.5 -> 3, 3.5 -> 4)."""
    if x >= 0:
        return int(x + 0.5)
    else:
        return int(x - 0.5)


# ==============================================================================
# SELF-TEST SUITE
# ==============================================================================
if __name__ == '__main__':
    # Test PIT-01
    assert demonstrate_recursion_fix() == 2500
    
    # Test PIT-02
    sd = SafeDict()
    sd[42] = "answer"
    assert sd[42] == "answer"
    assert 42 in sd
    assert sd.get(100, -1) == -1
    
    # Test PIT-03
    assert demonstrate_precision_pitfall() == 10**18
    assert math.isqrt(10**18) == 10**9
    
    # Test PIT-04
    assert len(fast_string_build(100)) == 100
    
    # Test PIT-05
    assert demonstrate_queue_performance() == 0
    
    # Test PIT-06
    mat = create_safe_matrix(3, 3, 0)
    mat[0][0] = 99
    assert mat[1][0] == 0, "Matrix rows must be independent!"
    
    # Test PIT-07
    assert is_even_bitwise(4) is True
    assert is_even_bitwise(5) is False
    assert mask_all_ones(3) == 7
    
    # Test PIT-08
    assert sort_numeric_strings(["100", "20", "5"]) == ["5", "20", "100"]
    
    # Test PIT-09
    assert remove_negatives_safely([-2, 5, -1, 3]) == [5, 3]
    
    # Test PIT-10
    assert running_window_sum([1, 2, 3, 4], 2) == [3, 5, 7]
    
    # Test PIT-11
    assert parse_all_integers("10 20  30\n40") == [10, 20, 30, 40]
    
    # Test PIT-12
    # Demonstrates distinction: round(2.5) == 2, but standard_round_half_up(2.5) == 3
    assert round(2.5) == 2
    assert standard_round_half_up(2.5) == 3
    assert standard_round_half_up(3.5) == 4

    print("ALL PitfallsAndErrors.py tests passed successfully!")

File: CORE/CPP/PitfallsAndErrors.cpp
/**
 * ================================================================================
 * 000_PitfallsAndErrors.cpp - C++ Competitive Programming Traps & Debugging Guide
 * ================================================================================
 * A comprehensive diagnostic handbook detailing subtle runtime traps, memory
 * leaks, undefined behavior (UB), and performance bottlenecks in C++ that cause
 * Time Limit Exceeded (TLE), Memory Limit Exceeded (MLE), Wrong Answer (WA),
 * and Runtime Errors (RTE / SIGSEGV) on Codeforces, LeetCode, and HackerRank.
 * 
 * Extracted from real issues encountered across 16+ HackerRank C++ problems.
 * 
 * TABLE OF CONTENTS:
 * --------------------------------------------------------------------------------
 * 1.  [PIT-01] The Leftover Newline Trap (Mixing scanf/cin >> with getline)
 * 2.  [PIT-02] Object Slicing in Exception Catch Blocks
 * 3.  [PIT-03] Memory Leaks & delete vs delete[] Mismatches
 * 4.  [PIT-04] Passing std::string to printf("%s") Without .c_str()
 * 5.  [PIT-05] Variable-Length Arrays (VLA) & Stack Overflow on Large N
 * 6.  [PIT-06] Signed vs Unsigned Integer Comparison (size_t vs int)
 * 7.  [PIT-07] Reusing std::stringstream Without Calling .clear()
 * 8.  [PIT-08] Buffer Desynchronization Hazard (Mixing cin with scanf)
 * ================================================================================
 */

#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <exception>
#include <stdexcept>
#include <cstdio>
#include <cassert>

// ==============================================================================
// [PIT-01] The Leftover Newline Trap (Mixing scanf/cin >> with getline)
// ==============================================================================
/**
 * Trap:
 *     Reading an integer with `scanf("%d", &n)` or `cin >> n` leaves the trailing
 *     newline character (`\n`) in the standard input stream buffer.
 *     When you immediately follow with `getline(cin, line)`, `getline` reads that
 *     residual newline and returns an EMPTY string instead of the intended text.
 * 
 * Why It Happens:
 *     Formatted extraction (`scanf`, `operator>>`) stops immediately when it reaches
 *     whitespace/newline, leaving the delimiter unconsumed in the buffer.
 *     `getline` terminates at the first `\n` it encounters.
 * 
 * How To Fix It:
 *     Consume the leftover newline before calling `getline`:
 *     Method A: `std::string dummy; std::getline(std::cin, dummy);`
 *     Method B: `std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');`
 * 
 * Repository Reference:
 *     - CPP/0004_Attribute-Parser.cpp (Lines 16-19: scanf("%d %d", &n, &q); getline(cin, temp);)
 */

std::pair<int, std::string> simulateInputWithNewlineTrap(const std::string& simulatedInput, bool applyFix) {
    std::stringstream stream(simulatedInput);
    int number;
    stream >> number; // Leaves '\n' in stream buffer
    
    std::string text;
    if (applyFix) {
        std::string dummy;
        std::getline(stream, dummy); // Discard residual newline
        std::getline(stream, text);  // Read actual line
    } else {
        std::getline(stream, text);  // Erroneously consumes residual newline!
    }
    return {number, text};
}


// ==============================================================================
// [PIT-02] Object Slicing in Exception Catch Blocks
// ==============================================================================
/**
 * Trap:
 *     Catching polymorphic exceptions by value:
 *         `catch (std::exception e)` // BUG: SLICED!
 *     instead of by reference:
 *         `catch (const std::exception& e)` // CORRECT
 * 
 * Why It Happens:
 *     Catching by value constructs a base `std::exception` copy of the thrown object,
 *     slicing off derived member variables and preventing virtual dispatch to
 *     the derived class's `what()` override.
 * 
 * How To Fix It:
 *     Always catch polymorphic exceptions by `const std::exception& e`.
 * 
 * Repository Reference:
 *     - CPP/0012_Exceptional-Server.cpp (Lines 48-52: catch (const std::bad_alloc& e) ... catch (const std::exception& e))
 */

class DerivedException : public std::exception {
private:
    std::string details;
public:
    DerivedException(std::string d) : details(d) {}
    const char* what() const noexcept override {
        return details.c_str();
    }
};

std::string demonstrateCatchByReference() {
    try {
        throw DerivedException("Custom derived error payload");
    } catch (const std::exception& e) {
        // By reference: virtual what() correctly resolves to DerivedException
        return e.what();
    }
}


// ==============================================================================
// [PIT-03] Memory Leaks & delete vs delete[] Mismatches
// ==============================================================================
/**
 * Trap:
 *     1. Allocating an array with `new Type[N]` but freeing with scalar `delete ptr;`
 *        instead of `delete[] ptr;`. This causes undefined behavior (destructors of
 *        subsequent elements are never invoked, and allocator heap metadata corrupts).
 *     2. Dropping pointer nodes in linked lists or trees without `delete`, leading to MLE.
 * 
 * How To Fix It:
 *     - Always match `new[]` with `delete[]`.
 *     - When removing linked list nodes, store `Node* temp = curr->next; curr->next = temp->next; delete temp;`.
 *     - In modern C++, prefer `std::vector` or `std::unique_ptr`.
 * 
 * Repository Reference:
 *     - CPP/0016_Classes-and-Objects.cpp (Lines 27, 40: Student *s = new Student[n]; ... delete[] s;)
 *     - 30 DAYS OF CODE/Day-24-More-Linked-Lists.cpp (Lines 27-29: delete temp;)
 */

void demonstrateArrayAllocationCleanup() {
    int* data = new int[100];
    for (int i = 0; i < 100; ++i) data[i] = i;
    // Correct array deallocation
    delete[] data;
}


// ==============================================================================
// [PIT-04] Passing std::string to printf("%s") Without .c_str()
// ==============================================================================
/**
 * Trap:
 *     Writing `printf("%s\n", myString);` where `myString` is a `std::string`.
 * 
 * Why It Happens:
 *     `printf` is a C function that expects a raw `const char*` pointer pointing to
 *     a null-terminated C string. Passing a C++ class object pushes internal class
 *     pointers and size fields onto the variadic stack, causing garbage output or SIGSEGV.
 * 
 * How To Fix It:
 *     Always call `.c_str()`:
 *     `printf("%s\n", myString.c_str());`
 * 
 * Repository Reference:
 *     - CPP/0004_Attribute-Parser.cpp (Line 61: printf("%s\n", db[query].c_str());)
 *     - CPP/0007_Strings.cpp (Line 18: printf("%s\n", c.c_str());)
 *     - CPP/0014_Structs.cpp (Line 26: printf("%s", st.first_name.c_str());)
 */

std::string safeFormatString(const std::string& input) {
    char buf[128];
    // .c_str() is strictly mandatory here
    std::snprintf(buf, sizeof(buf), "Value: %s", input.c_str());
    return std::string(buf);
}


// ==============================================================================
// [PIT-05] Variable-Length Arrays (VLA) & Stack Overflow on Large N
// ==============================================================================
/**
 * Trap:
 *     Using non-standard C-style Variable Length Arrays on the stack:
 *     `int arr[n];`
 *     where $N$ is taken from user input.
 * 
 * Why It Happens:
 *     Thread call stacks typically have a strict limit (1MB to 8MB).
 *     If $N = 10^6$ ints, that requires $\approx 4$ MB of stack space, causing
 *     instant stack overflow and SIGSEGV before execution even reaches your logic.
 * 
 * How To Fix It:
 *     Use `std::vector<int> arr(n);` (heap allocated) or global static storage.
 * 
 * Repository Reference:
 *     - CPP/0009_Arrays-Introduction.cpp (Shows int arr[n]; acceptable only for small N <= 1000)
 *     - CPP/0010_Variable-Sized-Arrays.cpp (Uses vector<vector<int>> for dynamic sizing)
 */

std::vector<int> allocateSafeVector(size_t n) {
    // Vector allocates on heap, immune to stack overflow
    return std::vector<int>(n, 0);
}


// ==============================================================================
// [PIT-06] Signed vs Unsigned Integer Comparison (size_t vs int)
// ==============================================================================
/**
 * Trap:
 *     Comparing signed `int` with unsigned `container.size()` (`size_t`):
 *     `int i = -1; if (i < v.size())` evaluates to FALSE!
 * 
 * Why It Happens:
 *     Under C++ integer promotion rules, when a signed int is compared with an
 *     unsigned type of the same or larger rank, the signed value is converted to
 *     unsigned. A negative value like `-1` becomes `18446744073709551615ULL`!
 * 
 * How To Fix It:
 *     - Use `size_t` for index variables: `for (size_t i = 0; i < v.size(); ++i)`
 *     - Or cast explicitly: `static_cast<int>(v.size())`
 * 
 * Repository Reference:
 *     - CPP/0004_Attribute-Parser.cpp (Line 44: for (size_t j = 0; j < tag_stack.size(); ++j))
 *     - CPP/0013_StringStream.cpp (Line 28: for (size_t i = 0; i < integers.size(); i++))
 */

bool demonstrateSignedUnsignedTrap() {
    std::vector<int> vec = {10, 20, 30};
    int negativeIndex = -1;
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wsign-compare"
    bool buggyComparison = (negativeIndex < vec.size()); // Evaluates to false!
#pragma GCC diagnostic pop
    bool safeComparison = (negativeIndex < static_cast<int>(vec.size())); // Evaluates to true!
    
    assert(!buggyComparison);
    assert(safeComparison);
    return safeComparison;
}


// ==============================================================================
// [PIT-07] Reusing std::stringstream Without Calling .clear()
// ==============================================================================
/**
 * Trap:
 *     Reusing a `std::stringstream` by setting `.str("new content")` after it hit EOF:
 *     All subsequent extractions `ss >> x` will SILENTLY FAIL!
 * 
 * Why It Happens:
 *     When a stream reaches the end of its buffer, it sets the `eofbit` and `failbit`.
 *     Calling `ss.str("new content")` replaces the internal buffer, but does NOT
 *     clear the error flags.
 * 
 * How To Fix It:
 *     Always call `ss.clear()` before reusing the stream:
 *     `ss.clear();`
 *     `ss.str("new content");`
 * 
 * Repository Reference:
 *     - CPP/0013_StringStream.cpp
 */

bool demonstrateStringStreamReset() {
    std::stringstream ss("100");
    int val1 = 0, val2 = 0;
    ss >> val1;
    assert(val1 == 100);
    assert(ss.eof()); // eofbit is now set

    // Attempt reuse without clear() -> FAILS!
    ss.str("200");
    ss >> val2;
    assert(val2 == 0); // Did not extract because eofbit was still active

    // Proper reuse with clear() -> SUCCEEDS!
    ss.clear(); // Resets eofbit/failbit to goodbit
    ss.str("300");
    ss >> val2;
    assert(val2 == 300);
    return true;
}


// ==============================================================================
// [PIT-08] Buffer Desynchronization Hazard (Mixing cin with scanf)
// ==============================================================================
/**
 * Trap:
 *     Calling `std::ios::sync_with_stdio(false);` and then mixing `cin` with `scanf`
 *     or `cout` with `printf`.
 * 
 * Why It Happens:
 *     `sync_with_stdio(false)` disconnects the C++ standard stream buffers from the
 *     underlying C stdio buffers to maximize performance.
 *     If you subsequently read from both, the two buffers read independent chunks
 *     from the operating system, interleaving and corrupting your input order!
 * 
 * How To Fix It:
 *     Once `std::ios::sync_with_stdio(false)` is enabled, use ONLY `cin` and `cout`.
 *     Never mix with `scanf` or `printf`.
 * 
 * Repository Reference:
 *     - CPP/0015_Pointer.cpp (Lines 10-11: uses sync_with_stdio(false) with pure cin/cout)
 */


// ==============================================================================
// SELF-TEST VERIFICATION SUITE
// ==============================================================================
int main() {
    std::cout << "[RUNNING] PitfallsAndErrors.cpp Self-Test Suite...\n";

    // Test PIT-01: Newline Trap Simulation
    std::string testInput = "42\nHelloWorld\n";
    auto trapped = simulateInputWithNewlineTrap(testInput, false);
    assert(trapped.first == 42);
    assert(trapped.second == ""); // Trapped: consumed empty newline!

    auto fixed = simulateInputWithNewlineTrap(testInput, true);
    assert(fixed.first == 42);
    assert(fixed.second == "HelloWorld"); // Fixed!

    // Test PIT-02: Catch by Reference Slicing Prevention
    assert(demonstrateCatchByReference() == "Custom derived error payload");

    // Test PIT-03: Array Cleanup
    demonstrateArrayAllocationCleanup();

    // Test PIT-04: .c_str() Safe Formatting
    assert(safeFormatString("Antigravity") == "Value: Antigravity");

    // Test PIT-05: Heap Vector
    auto safeVec = allocateSafeVector(1000);
    assert(safeVec.size() == 1000);

    // Test PIT-06: Signed vs Unsigned Comparison
    assert(demonstrateSignedUnsignedTrap() == true);

    // Test PIT-07: StringStream Reset
    assert(demonstrateStringStreamReset() == true);

    std::cout << "[SUCCESS] ALL PitfallsAndErrors.cpp tests passed cleanly!\n";
    return 0;
}
